op-node/l1/batching.go

```diff
@@ -2,134 +2,103 @@ package l1
-				completed <- batchResult{failed: failed, err: err, success: len(batch) - len(failed)}
-			}
-		}(ctx)
+	return &IterativeBatchCall{
+		completed: 0,
+		getBatch:  getBatch,
+		requests:  requests,
+		scheduled: scheduled,
 	}
+}
 
-	parallelRequests := func() int {
-		// we split the requests into parallel batch requests, and count how many
-		i := 0
-		for ; i < maxParallel && len(requests) > 0; i++ {
-			nextBatch := requests
-			if len(nextBatch) > maxPerBatch {
-				nextBatch = requests[:maxPerBatch]
-			}
-			// don't retry this batch of requests again, unless we add them back
-			requests = requests[len(nextBatch):]
-
-			// schedule the batch, this may block if all workers are busy and the queue is full
-			batchRequests <- nextBatch
+// Fetch fetches more of the data, and returns io.EOF when all data has been fetched.
+// This method is safe to call concurrently: it will parallelize the fetching work.
+// If no work is available, but the fetching is not done yet,
+// then Fetch will block until the next thing can be fetched, or until the context expires.
+func (ibc *IterativeBatchCall) Fetch(ctx context.Context, maxBatchSize uint) error {
+	// collect a batch from the requests channel
+	batch := make([]rpc.BatchElem, 0, maxBatchSize)
+	// wait for first element
+	select {
+	case reqElem, ok := <-ibc.scheduled:
+		if !ok { // no more requests to do
+			return io.EOF
 		}
-		return i
+		batch = append(batch, reqElem)
+	case <-ctx.Done():
+		return ctx.Err()
 	}
 
-	maxCount := expectedBatches * maxRetry
-
-	awaited := len(requests)
-
-	// start initial round of parallel requests
-	count := parallelRequests()
-
-	// We slow down additional batch requests to not spam the server.
-	retryTicker := time.NewTicker(time.Millisecond * 20)
-	defer retryTicker.Stop()
-
-	// The main requests slice is only ever mutated by the go-routine running this loop.
-	// Slices of this are sent to worker go-routines, and never overwritten with different requests.
+	// collect more elements, if there are any.
 	for {
-		// check if we've all results back successfully
-		if awaited <= 0 {
-			return nil
-		}
-		if count > maxCount {
-			return TooManyRetries
+		if uint(len(batch)) >= maxBatchSize {
+			break
 		}
 		select {
-		case <-retryTicker.C:
-			count += parallelRequests() // retry batch-requests on interval
-		case result := <-completed:
-			if result.err != nil {
-				// batch failed, RPC may be broken, abort
-				return fmt.Errorf("batch request failed: %w", result.err)
+		case reqElem, ok := <-ibc.scheduled:
+			if !ok { // no more requests to do
+				return io.EOF
 			}
-			// if any element failed, add it to the requests for re-attempt
-			requests = append(requests, result.failed...)
-			awaited -= result.success
+			batch = append(batch, reqElem)
+			continue
 		case <-ctx.Done():
 			return ctx.Err()
+		default:
+			break
 		}
+		break
 	}
+
+	if err := ibc.getBatch(ctx, batch); err != nil {
+		for _, r := range batch {
```

Comment:
I see that we reschedule the entire batch here, would it be too complex to use multierror to only reschedule the requests that need to be rescheduled? Maybe I'm making some assumptions that aren't correct in this context

Reply:
This is a global error, it only happens if the get_batch RPC as a whole fails, so all items are always all failing, and have to be put back in the scheduling channel. See below for the per-item error handling, where only the failing items are put back into the scheduling channel.